Sources/HummingbirdCompression/RequestDecompressionMiddleware.swift

```diff
@@ -0,0 +1,111 @@
+    public init(windowSize: Int = 32768) {
+        self.windowSize = windowSize
+    }
+
+    public func handle(_ request: Request, context: Context, next: (Request, Context) async throws -> Response) async throws -> Response {
+        if let algorithm = algorithm(from: request.headers[values: .contentEncoding]) {
+            var request = request
+            request.body = .init(asyncSequence: DecompressByteBufferSequence(
+                base: request.body,
+                algorithm: algorithm,
+                windowSize: self.windowSize
+            ))
+            let response = try await next(request, context)
+            return response
+        } else {
+            return try await next(request, context)
+        }
+    }
+
+    /// Determines the decompression algorithm based off content encoding header.
+    private func algorithm(from contentEncodingHeaders: [String]) -> CompressionAlgorithm? {
+        for encoding in contentEncodingHeaders {
+            switch encoding {
+            case "gzip":
+                return CompressionAlgorithm.gzip()
+            case "deflate":
+                return CompressionAlgorithm.zlib()
+            default:
+                break
+            }
+        }
+        return nil
+    }
+}
+
+/// AsyncSequence of decompressed ByteBuffers
+struct DecompressByteBufferSequence<Base: AsyncSequence & Sendable>: AsyncSequence, Sendable where Base.Element == ByteBuffer {
+    typealias Element = ByteBuffer
+
+    let base: Base
+    let algorithm: CompressionAlgorithm
+    let windowSize: Int
+
+    class AsyncIterator: AsyncIteratorProtocol {
+        var baseIterator: Base.AsyncIterator
+        let decompressor: NIODecompressor
+        var currentBuffer: ByteBuffer?
+        var window: ByteBuffer
+
+        init(baseIterator: Base.AsyncIterator, algorithm: CompressionAlgorithm, windowSize: Int) {
+            self.baseIterator = baseIterator
+            self.decompressor = algorithm.decompressor
+            self.window = ByteBufferAllocator().buffer(capacity: windowSize)
+            self.currentBuffer = nil
+            try! self.decompressor.startStream()
+        }
+
+        deinit {
+            try! self.decompressor.finishStream()
+        }
+
+        func next() async throws -> ByteBuffer? {
+            if self.currentBuffer == nil {
+                self.currentBuffer = try await self.baseIterator.next()
+            }
+            self.window.moveReaderIndex(to: 0)
+            self.window.moveWriterIndex(to: 0)
+            while var buffer = self.currentBuffer {
+                do {
+                    try buffer.decompressStream(to: &self.window, with: self.decompressor)
+                } catch let error as CompressNIOError where error == .bufferOverflow {
+                    self.currentBuffer = buffer
+                    return self.window
+                } catch let error as CompressNIOError where error == .inputBufferOverflow {
```

Comment:
`case CompressNIOError.inputBufferOverflow `

Reply:
Error isn't an enum so this won't work
I could implement operator ~= in CompressNIO I guess